### PR TITLE
Do not re-check visibility or re-render shadow maps for point and spot lights for each view

### PR DESCRIPTION
**Pull request by coreh**
# Objective

_If I understand it correctly_, we were checking mesh visibility, as well as re-rendering point and spot light shadow maps for each view. This makes it so that M views and N lights produce M x N complexity. This PR aims to fix that, as well as introduce a stress test for this specific scenario.

## Solution

- Keep track of what lights have already had mesh visibility calculated and do not calculate it again;
- Reuse shadow depth textures and attachments across all views, and only render shadow maps for the _first_ time a light is encountered on a view;
- Directional lights remain unaltered, since their shadow map cascades are view-dependent;
- Add a new `many_cameras_lights` stress test example to verify the solution

## Showcase

110% speed up on the stress test
83% reduction of memory usage in stress test

### Before (5.35 FPS on stress test)
Screenshot 2024-09-11 at 12 25 57

### After (11.34 FPS on stress test)
Screenshot 2024-09-11 at 12 24 35

## Testing

- Did you test these changes? If so, how? 
  - On my game project where I have two cameras, and many shadow casting lights I managed to get pretty much double the FPS.
  - Also included a stress test, see the comparison above
- Are there any parts that need more testing?
  - Yes, I would like help verifying that this fix is indeed correct, and that we were really re-rendering the shadow maps by mistake and it's indeed okay to not do that
- How can other people (reviewers) test your changes? Is there anything specific they need to know?
  - Run the `many_cameras_lights` example
  - On the `main` branch, cherry pick the commit with the example (`git cherry-pick --no-commit 1ed4ace01`) and run it
- If relevant, what platforms did you test these changes on, and are there any important ones you can't test?
  - macOS

### CODE REVIEW

**Review by IceSentry — APPROVED** [review]

**Review by JMS55 — APPROVED** [review]
Kinda cursed, but I'm fine merging it. The whole renderer needs a cleanup around resource creation and handling, one more cludge won't hurt 😅.

**Comment by alice-i-cecile** [comment]
@coreh, are you able to resolve the merge conflicts here? :)

**Comment by BenjaminBrienen** [comment]
I'll take a look at this tomorrow and see if the merge conflicts are simple enough.

**Comment by alice-i-cecile** [comment]
I can't merge this: CI is failing due to disallowed platform-inconsistent float methods. It's an easy fix at least :)

File: examples/stress_tests/many_cameras_lights.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
+//! Test rendering of many cameras and lights
+
+use std::f32::consts::PI;
+
+use bevy::prelude::*;
+use bevy_render::camera::Viewport;
```
Comment (mockersf) [inline examples/stress_tests/many_cameras_lights.rs:6]:
```suggestion
use bevy::{
    math::ops::{cos, sin},
    prelude::*,
    render::camera::Viewport,
};
```

File: examples/stress_tests/many_cameras_lights.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    for i in 0..NUM_LIGHTS {
+        let angle = (i as f32) / (NUM_LIGHTS as f32) * PI * 2.0;
+        commands.spawn((
+            PointLight {
+                color: Color::hsv(angle.to_degrees(), 1.0, 1.0),
+                intensity: 2_000_000.0 / NUM_LIGHTS as f32,
+                shadows_enabled: true,
+                ..default()
+            },
+            Transform::from_xyz(angle.sin() * 4.0, 2.0, angle.cos() * 4.0),
```
Comment (mockersf) [inline examples/stress_tests/many_cameras_lights.rs:51]:
```suggestion
            Transform::from_xyz(sin(angle) * 4.0, 2.0, cos(angle) * 4.0),
```

File: examples/stress_tests/many_cameras_lights.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                            (x as f32 * width) as u32,
+                            (y as f32 * height) as u32,
+                        ),
+                        physical_size: UVec2::new(width as u32, height as u32),
+                        ..default()
+                    }),
+                    order: i,
+                    ..default()
+                },
+                Transform::from_xyz(angle.sin() * 4.0, 2.5, angle.cos() * 4.0)
```
Comment (mockersf) [inline examples/stress_tests/many_cameras_lights.rs:77]:
```suggestion
                Transform::from_xyz(sin(angle) * 4.0, 2.5, cos(angle) * 4.0)
```

**Pull request merged**